File: dotnet/Common/TicketLoader.cs
using System.Text.Json;
using System.Text.Json.Serialization;
using Microsoft.Extensions.Configuration;

namespace WorkflowLab.Common;

/// <summary>
/// Utility class for loading tickets from the external JSON data file.
/// Supports loading all tickets, querying by ID, or selecting a random ticket.
/// The tickets path must be configured in appsettings.Local.json via TICKETS_PATH.
/// </summary>
public static class TicketLoader
{
    private static IConfiguration? _configuration;
    private static string? _configPath;

    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNameCaseInsensitive = true,
        Converters = { new JsonStringEnumConverter(JsonNamingPolicy.CamelCase) }
    };

    /// <summary>
    /// Gets the configuration, loading from appsettings.Local.json in the dotnet folder.
    /// </summary>
    private static IConfiguration Configuration
    {
        get
        {
            if (_configuration == null)
            {
                _configPath = FindConfigPath(AppContext.BaseDirectory);
                _configuration = new ConfigurationBuilder()
                    .SetBasePath(_configPath)
                    .AddJsonFile("appsettings.Local.json", optional: true, reloadOnChange: true)
                    .AddEnvironmentVariables()
                    .Build();
            }
            return _configuration;
        }
    }

    /// <summary>
    /// Finds the dotnet folder by traversing up the directory tree.
    /// </summary>
    private static string FindConfigPath(string startPath)
    {
        var currentDir = new DirectoryInfo(startPath);
        while (currentDir != null)
        {
            if (currentDir.Name.Equals("dotnet", StringComparison.OrdinalIgnoreCase))
            {
                return currentDir.FullName;
            }
            currentDir = currentDir.Parent;
        }
        return startPath;
    }

    /// <summary>
    /// Gets the configured tickets path from TICKETS_PATH in appsettings.Local.json.
    /// </summary>
    private static string GetTicketsPath()
    {
        var ticketsPath = Configuration["TICKETS_PATH"];
        
        if (string.IsNullOrWhiteSpace(ticketsPath))
        {
            throw new InvalidOperationException(
                "TICKETS_PATH is not configured. " +
                "Set 'TICKETS_PATH' in appsettings.Local.json or as an environment variable. " +
                "Example: \"TICKETS_PATH\": \"..\\\\data\\\\tickets.json\"");
        }

        // Resolve relative paths from the config directory
        if (!Path.IsPathRooted(ticketsPath))
        {
            ticketsPath = Path.Combine(_configPath ?? AppContext.BaseDirectory, ticketsPath);
        }

        return Path.GetFullPath(ticketsPath);
    }

    /// <summary>
    /// Loads all tickets from the JSON file.
    /// </summary>
    public static async Task<List<SupportTicket>> LoadAllTicketsAsync()
    {
        var resolvedPath = GetTicketsPath();

        if (!File.Exists(resolvedPath))
        {
            throw new FileNotFoundException($"Tickets file not found at: {resolvedPath}");
        }

        var json = await File.ReadAllTextAsync(resolvedPath);
        var ticketDtos = JsonSerializer.Deserialize<List<TicketDto>>(json, JsonOptions)
            ?? throw new InvalidOperationException("Failed to deserialize tickets.");

        return ticketDtos.Select(MapToSupportTicket).ToList();
    }

    /// <summary>
    /// Gets a ticket by its ID.
    /// </summary>
    public static async Task<SupportTicket?> GetTicketByIdAsync(string ticketId)
    {
        var tickets = await LoadAllTicketsAsync();
        return tickets.FirstOrDefault(t => 
            t.TicketId.Equals(ticketId, StringComparison.OrdinalIgnoreCase));
    }

    /// <summary>
    /// Gets a random ticket from the collection.
    /// </summary>
    public static async Task<SupportTicket> GetRandomTicketAsync()
    {
        var tickets = await LoadAllTicketsAsync();
        if (tickets.Count == 0)
        {
            throw new InvalidOperationException("No tickets found in the data file.");
        }

        var random = new Random();
        return tickets[random.Next(tickets.Count)];
    }

    /// <summary>
    /// Gets a ticket by index (1-based for user friendliness).
    /// </summary>
    public static async Task<SupportTicket> GetTicketByIndexAsync(int index)
    {
        var tickets = await LoadAllTicketsAsync();
        if (index < 1 || index > tickets.Count)
        {
            throw new ArgumentOutOfRangeException(nameof(index), 
                $"Index must be between 1 and {tickets.Count}.");
        }

        return tickets[index - 1];
    }

    /// <summary>
    /// Lists all available tickets (for display purposes).
    /// </summary>
    public static async Task DisplayAvailableTicketsAsync()
    {
        var tickets = await LoadAllTicketsAsync();
        Console.WriteLine("Available tickets:");
        Console.WriteLine(new string('-', 60));
        for (int i = 0; i < tickets.Count; i++)
        {
            var t = tickets[i];
            Console.WriteLine($"  [{i + 1}] {t.TicketId} - {t.Subject} ({t.Priority})");
        }
        Console.WriteLine(new string('-', 60));
    }

    private static SupportTicket MapToSupportTicket(TicketDto dto)
    {
        var priority = dto.Priority?.ToUpperInvariant() switch
        {
            "LOW" => TicketPriority.Low,
            "MEDIUM" => TicketPriority.Medium,
            "HIGH" => TicketPriority.High,
            "CRITICAL" => TicketPriority.Critical,
            _ => TicketPriority.Medium
        };

        return new SupportTicket(
            TicketId: dto.Id ?? "UNKNOWN",
            CustomerId: dto.CustomerId ?? "UNKNOWN",
            CustomerName: dto.CustomerName ?? "Unknown Customer",
            Subject: dto.Subject ?? "No Subject",
            Description: dto.Description ?? "No Description",
            Priority: priority
        );
    }

    /// <summary>
    /// Internal DTO for JSON deserialization.
    /// </summary>
    private sealed record TicketDto(
        [property: JsonPropertyName("id")] string? Id,
        [property: JsonPropertyName("customerId")] string? CustomerId,
        [property: JsonPropertyName("customerName")] string? CustomerName,
        [property: JsonPropertyName("subject")] string? Subject,
        [property: JsonPropertyName("description")] string? Description,
        [property: JsonPropertyName("status")] string? Status,
        [property: JsonPropertyName("priority")] string? Priority,
        [property: JsonPropertyName("assignedTo")] string? AssignedTo
    );
}

File: python/common/ticket_loader.py
"""
Ticket Loader

Utility for loading tickets from the external JSON data file.
Supports loading all tickets, querying by ID, or selecting a random ticket.
The tickets path must be configured in .env via TICKETS_PATH.
"""

import json
import os
import random
from pathlib import Path
from typing import Optional

from .support_ticket import SupportTicket, TicketPriority


def _get_tickets_path() -> Path:
    """
    Get the configured tickets path from TICKETS_PATH environment variable.
    
    Returns:
        Path to the tickets JSON file.
        
    Raises:
        ValueError: If TICKETS_PATH is not configured.
    """
    tickets_path = os.environ.get("TICKETS_PATH")
    
    if not tickets_path:
        raise ValueError(
            "TICKETS_PATH is not configured. "
            "Set 'TICKETS_PATH' in .env or as an environment variable. "
            'Example: "TICKETS_PATH": "../data/tickets.json"'
        )
    
    path = Path(tickets_path)
    
    # Resolve relative paths from the python folder
    if not path.is_absolute():
        python_folder = Path(__file__).parent.parent
        path = python_folder / path
    
    return path.resolve()


def _map_priority(priority_str: Optional[str]) -> TicketPriority:
    """Map priority string to TicketPriority enum."""
    if not priority_str:
        return TicketPriority.MEDIUM
    
    priority_upper = priority_str.upper()
    mapping = {
        "LOW": TicketPriority.LOW,
        "MEDIUM": TicketPriority.MEDIUM,
        "HIGH": TicketPriority.HIGH,
        "CRITICAL": TicketPriority.CRITICAL,
    }
    return mapping.get(priority_upper, TicketPriority.MEDIUM)


def _map_to_support_ticket(dto: dict) -> SupportTicket:
    """Map a dictionary to a SupportTicket."""
    return SupportTicket(
        ticket_id=dto.get("id", "UNKNOWN"),
        customer_id=dto.get("customerId", "UNKNOWN"),
        customer_name=dto.get("customerName", "Unknown Customer"),
        subject=dto.get("subject", "No Subject"),
        description=dto.get("description", "No Description"),
        priority=_map_priority(dto.get("priority")),
    )


def load_all_tickets() -> list[SupportTicket]:
    """
    Load all tickets from the JSON file.
    
    Returns:
        List of SupportTicket objects.
        
    Raises:
        ValueError: If TICKETS_PATH is not configured.
        FileNotFoundError: If the tickets file is not found.
    """
    resolved_path = _get_tickets_path()
    
    if not resolved_path.exists():
        raise FileNotFoundError(f"Tickets file not found at: {resolved_path}")
    
    with open(resolved_path, "r", encoding="utf-8") as f:
        ticket_dtos = json.load(f)
    
    return [_map_to_support_ticket(dto) for dto in ticket_dtos]


def get_ticket_by_id(ticket_id: str) -> Optional[SupportTicket]:
    """
    Get a ticket by its ID.
    
    Args:
        ticket_id: The ticket ID to search for.
        
    Returns:
        The matching SupportTicket or None if not found.
    """
    tickets = load_all_tickets()
    for ticket in tickets:
        if ticket.ticket_id.lower() == ticket_id.lower():
            return ticket
    return None


def get_random_ticket() -> SupportTicket:
    """
    Get a random ticket from the collection.
        
    Returns:
        A randomly selected SupportTicket.
        
    Raises:
        ValueError: If no tickets are found.
    """
    tickets = load_all_tickets()
    if not tickets:
        raise ValueError("No tickets found in the data file.")
    
    return random.choice(tickets)


def get_ticket_by_index(index: int) -> SupportTicket:
    """
    Get a ticket by index (1-based for user friendliness).
    
    Args:
        index: 1-based index of the ticket.
        
    Returns:
        The SupportTicket at the given index.
        
    Raises:
        IndexError: If the index is out of range.
    """
    tickets = load_all_tickets()
    if index < 1 or index > len(tickets):
        raise IndexError(f"Index must be between 1 and {len(tickets)}.")
    
    return tickets[index - 1]


def display_available_tickets() -> None:
    """
    Display all available tickets for selection.
    """
    tickets = load_all_tickets()
    print("Available tickets:")
    print("-" * 60)
    for i, t in enumerate(tickets, 1):
        print(f"  [{i}] {t.ticket_id} - {t.subject} ({t.priority.value})")
    print("-" * 60)
